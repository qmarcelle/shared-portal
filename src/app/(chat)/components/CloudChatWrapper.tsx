'use client';

// CloudChatWrapper injects the Genesys Cloud Messenger SDK and manages Messenger-specific events.
// All script loading, Messenger events, and errors are logged for traceability and debugging.

import { useEffect } from 'react';
import { useChatStore } from '../stores/chatStore';

export default function CloudChatWrapper() {
  const { userData } = useChatStore();

  useEffect(() => {
    // Build chatSettings from env and userData
    window.chatSettings = {
      bootstrapUrl: process.env.NEXT_PUBLIC_GENESYS_BOOTSTRAP_URL || '',
      widgetUrl: process.env.NEXT_PUBLIC_GENESYS_WIDGET_URL || '',
      clickToChatJs: process.env.NEXT_PUBLIC_GENESYS_CLICK_TO_CHAT_JS || '',
      clickToChatEndpoint: process.env.NEXT_PUBLIC_CLICK_TO_CHAT_ENDPOINT || '',
      chatTokenEndpoint: process.env.NEXT_PUBLIC_CHAT_TOKEN_ENDPOINT || '',
      coBrowseEndpoint: process.env.NEXT_PUBLIC_COBROWSE_LICENSE_ENDPOINT || '',
      opsPhone: process.env.NEXT_PUBLIC_OPS_PHONE || '',
      opsPhoneHours: process.env.NEXT_PUBLIC_OPS_HOURS || '',
      ...userData,
    };
    console.debug('[CloudChatWrapper] chatSettings:', window.chatSettings);
  }, [userData]);

  useEffect(() => {
    // Load Genesys Cloud bootstrap and widget scripts
    const bootstrapUrl = window.chatSettings?.bootstrapUrl || '';
    const widgetUrl = window.chatSettings?.widgetUrl || '';
    if (!bootstrapUrl) return;
    const bootstrapScript = document.createElement('script');
    bootstrapScript.src = bootstrapUrl;
    bootstrapScript.async = true;
    bootstrapScript.onload = () => {
      console.debug('[CloudChatWrapper] Genesys bootstrap loaded');
      if (!widgetUrl) return;
      const widgetScript = document.createElement('script');
      widgetScript.src = widgetUrl;
      widgetScript.async = true;
      widgetScript.onload = () => {
        console.debug('[CloudChatWrapper] Genesys widgets loaded');
      };
      widgetScript.onerror = (e) => {
        console.error('[CloudChatWrapper] Failed to load widgets script', e);
      };
      document.body.appendChild(widgetScript);
    };
    bootstrapScript.onerror = (e) => {
      console.error('[CloudChatWrapper] Failed to load bootstrap script', e);
    };
    document.body.appendChild(bootstrapScript);

    // Provide openGenesysChat logic
    window.openGenesysChat = () => {
      if (window.Genesys) {
        try {
          window.Genesys('command', 'Messenger.open');
        } catch (e) {
          console.error(
            '[CloudChatWrapper] Error opening Genesys Messenger',
            e,
          );
        }
      }
    };

    return () => {
      // Cleanup scripts and chatSettings
      document.body.removeChild(bootstrapScript);
      if (widgetUrl) {
        const widgetScript = document.querySelector(
          'script[src="' + widgetUrl + '"]',
        );
        if (widgetScript) document.body.removeChild(widgetScript);
      }
      if (window.chatSettings) delete window.chatSettings;
      if (window.openGenesysChat) delete window.openGenesysChat;
      console.debug('[CloudChatWrapper] Cleanup complete');
    };
  }, [userData]);

  return null;
}
