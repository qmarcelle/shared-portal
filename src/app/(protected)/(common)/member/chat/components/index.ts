/**
 * Component Exports
 * Centralizes all component exports following the new consolidated structure
 */

// Main Components
export { ChatTrigger } from './ChatTrigger';
export { ChatWidget } from './ChatWidget';
export { PlanInfoHeader } from './PlanInfoHeader';

// Error Handling
export { ChatErrorBoundary } from './shared/ChatErrorBoundary';

// Integration Components (to be consolidated)
export { GenesysScripts } from './GenesysScripts';
