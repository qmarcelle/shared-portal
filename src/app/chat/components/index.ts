// Core Components
export * from './core/ChatButton';
export * from './core/ChatErrorBoundary';
export * from './core/ChatHeader';
export * from './core/ChatWidget';

// Business Hours Components
export * from './business-hours/BusinessHoursNotification';

// Eligibility Components
export * from './eligibility/EligibilityCheck';

// Plan Switcher Components will be added here when implemented
// export * from './plan-switcher/PlanSwitcher';
