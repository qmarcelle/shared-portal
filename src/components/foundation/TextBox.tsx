import { IComponent } from '../IComponent';

interface TextBoxProps extends IComponent {
  text: string;
  tabFocus?: number;
  display?: 'block' | 'inline';
  type?: 'title-1' | 'title-2' | 'title-3' | 'body-1' | 'body-2';
}

export const TextBox = ({
  text,
  type = 'body-1',
  tabFocus,
  className = '',
  display = 'block',
}: TextBoxProps) => {
  return (
    <p
      className={`${type} ${className}`.trimEnd()}
      tabIndex={tabFocus}
      style={{ display: display }}
    >
      {text}
    </p>
  );
};
