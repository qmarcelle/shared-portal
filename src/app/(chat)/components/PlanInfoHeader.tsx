'use client';
import { useChatStore } from '@/app/(chat)/stores/chatStore';
import { Title } from '@/components/foundation/Title';

/**
 * Show current plan when multiple plans exist
 */
export default function PlanInfoHeader() {
  const { formInputs } = useChatStore();
  if (formInputs.length <= 1) return null;
  const plan = formInputs.find((f) => f.id === 'PLAN_ID');

  return (
    <Title
      className="plan-info-header"
      text={`You are chatting about plan ${plan?.value || ''}`}
    />
  );
}
