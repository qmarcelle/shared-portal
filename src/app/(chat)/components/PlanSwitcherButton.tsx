'use client';
import { useChatStore } from '@/app/(chat)/stores/chatStore';
import { Button } from '@/components/foundation/Button';
import { usePlanStore } from '../../../userManagement/stores/planStore';

/**
 * "Switch Plan" inside pre-chat window
 */
export default function PlanSwitcherButton() {
  const { endChat } = useChatStore();
  const { plans, selectedPlanId, openPlanSwitcher } = usePlanStore();

  if (plans.length <= 1) return null;

  const handleSwitch = () => {
    endChat();
    openPlanSwitcher();
  };

  return (
    <Button
      label="Switch Plan"
      callback={handleSwitch}
      className="plan-switch-button"
      type="secondary"
    />
  );
}
