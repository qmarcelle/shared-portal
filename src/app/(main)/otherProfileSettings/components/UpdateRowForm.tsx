import Image from 'next/image';
import { ReactElement, useState } from 'react';
import editIcon from '../../../../../public/assets/edit.svg';
import { IComponent } from '../../../../components/IComponent';
import { Button } from '../../../../components/foundation/Button';
import { Checkbox } from '../../../../components/foundation/Checkbox';
import { Column } from '../../../../components/foundation/Column';
import { Divider } from '../../../../components/foundation/Divider';
import { Radio } from '../../../../components/foundation/Radio';
import { Row } from '../../../../components/foundation/Row';
import { Spacer } from '../../../../components/foundation/Spacer';
import { TextBox } from '../../../../components/foundation/TextBox';
import { TextField } from '../../../../components/foundation/TextField';
import { Title } from '../../../../components/foundation/Title';
type OptionType = 'radio' | 'checkbox' | 'textbox' | 'currentselection';

export interface UpdateRowFormProps extends IComponent {
  label: ReactElement;
  subLabel?: string;
  divider?: boolean;
  enabled: boolean;
  optionObjects: OptionData[];
  saveCallback?: () => void | Promise<void> | null;
  cancelCallback?: boolean;
  type: OptionType;
  icon?: JSX.Element;
}

export interface OptionData {
  label: string;
  enabled: boolean;
}

export const UpdateRowForm = ({
  label,
  subLabel,
  divider = false,
  type,
  optionObjects,
  saveCallback,
  cancelCallback,
}: UpdateRowFormProps) => {
  const [isCancel, setCancel] = useState(false);

  const updateCallBack = () => {
    setCancel(true);
  };

  return (
    <Column className="w-[460px]">
      {label}
      <Column className="body-2">{subLabel}</Column>
      <Spacer size={16} />
      <Column>
        {type === 'radio' &&
          optionObjects.map((OptionData, index) => (
            <Column key={index}>
              <Radio label={OptionData.label} selected={OptionData.enabled} />
            </Column>
          ))}
        {type === 'checkbox' &&
          optionObjects.map((OptionData, index) => (
            <Column key={index}>
              <Checkbox label={OptionData.label} selected={false} />
            </Column>
          ))}
        {type === 'textbox' && (
          <Column className=" w-[300px] body-1 ">
            <Radio label="Enter Language" selected={false} />
            <Column className="px-10 ">
              <TextField type="text" label=""></TextField>
            </Column>
            <Radio label="Decline to answer" selected={false} />
          </Column>
        )}
        {type === 'currentselection' && (
          <Column className=" body-1 px-1" onClick={saveCallback}>
            <TextBox text="Current Selection: Hispanic or Latino" />
            <Spacer size={8} />
            <Title
              className="font-bold primary-color"
              text="Update"
              suffix={<Image src={editIcon} alt="link" />}
            />
          </Column>
        )}
      </Column>
      <Spacer size={16} />
      <Row className="flex-row">
        {type != 'currentselection' && (
          <Button
            className="w-[250px]"
            label="Save Answer"
            type="primary"
            callback={saveCallback}
          />
        )}
        <Spacer axis="horizontal" size={16} />
        {cancelCallback === true && isCancel === false && (
          <Button
            className="w-[250px]"
            label="Cancel"
            type="secondary"
            callback={updateCallBack}
          />
        )}
      </Row>
      <Spacer size={32} />
      {divider && <Divider />}
    </Column>
  );
};
