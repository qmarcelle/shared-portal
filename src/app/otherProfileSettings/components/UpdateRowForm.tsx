import { IComponent } from '@/components/IComponent';
import { Button } from '@/components/foundation/Button';
import { Checkbox } from '@/components/foundation/Checkbox';
import { Column } from '@/components/foundation/Column';
import { Divider } from '@/components/foundation/Divider';
import { Radio } from '@/components/foundation/Radio';
import { Row } from '@/components/foundation/Row';
import { Spacer } from '@/components/foundation/Spacer';
import { TextBox } from '@/components/foundation/TextBox';
import { TextField } from '@/components/foundation/TextField';
import { Title } from '@/components/foundation/Title';
import Image from 'next/image';
import { ReactElement } from 'react';
import editIcon from '../../../../public/assets/edit.svg';
import { LangauageDetails } from '../model/api/language';
type OptionType = 'radio' | 'checkbox' | 'textbox' | 'currentselection';

export interface UpdateRowFormProps extends IComponent {
  label: ReactElement;
  subLabel?: string;
  divider?: boolean;
  enabled: boolean;
  optionObjects: OptionData[];
  saveCallback?: () => void | Promise<void> | null;
  cancelCallback?: () => void;
  type: OptionType;
  icon?: JSX.Element;
  optionsEnabled?: boolean;
  isRaceField?: boolean;
  languageOptions?: LangauageDetails[];
  isDeclineLanguageSelected?: boolean;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  selectionCallBack?: (val: any, isChecked?: boolean) => void;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  languageSelectionCallBack?: (val: any) => void;
  validLanguage?: boolean;
  languageEmptySelect?: boolean;
  isImpersonated?: boolean;
}

export interface OptionData {
  label: string;
  code: string;
  enabled: boolean;
}

export const UpdateRowForm = ({
  label,
  subLabel,
  divider = false,
  type,
  optionObjects,
  saveCallback,
  cancelCallback,
  isRaceField,
  optionsEnabled,
  languageOptions,
  isDeclineLanguageSelected,
  selectionCallBack,
  languageSelectionCallBack,
  validLanguage,
  languageEmptySelect,
  isImpersonated = false,
}: UpdateRowFormProps) => {
  function currentSelectionText(optionObjects: OptionData[]) {
    if (!isRaceField)
      return <TextBox text={'Current Selection: ' + optionObjects[0].label} />;
    else {
      const raceText = optionObjects.map((item) => item.label).join(',');
      return <TextBox text={`Current Selection: ${raceText}`} />;
    }
  }

  function languageErrorText() {
    if (!validLanguage) return ['Invalid entry.Please try again'];
    else return [];
  }

  function checkImpersonationForSaveCallback() {
    if (isImpersonated) {
      alert('This functionality is not available in impersonation mode');
      return;
    }
    saveCallback?.();
  }

  return (
    <Column className="w-[460px]">
      {label}
      <Column className="body-2">{subLabel}</Column>
      <Spacer size={16} />
      <Column>
        {type === 'radio' &&
          optionObjects.map((OptionData, index) => (
            <Column key={index}>
              <Radio
                label={OptionData.label}
                selected={OptionData.enabled}
                value={OptionData.code}
                callback={(val) => selectionCallBack?.(val)}
              />
            </Column>
          ))}
        {type === 'checkbox' &&
          optionObjects.map((OptionData, index) => (
            <Column key={index}>
              <Checkbox
                label={OptionData.label}
                checked={OptionData.enabled}
                onChange={(isChecked) =>
                  selectionCallBack?.(OptionData.code, isChecked)
                }
              />
            </Column>
          ))}
        {type === 'textbox' && (
          <div className=" w-[300px] body-1">
            <Radio
              label="Enter Language"
              selected={
                !isDeclineLanguageSelected && Boolean(languageEmptySelect)
              }
              value="L2"
              callback={(val) => selectionCallBack?.(val)}
            />
            {!isDeclineLanguageSelected && Boolean(languageEmptySelect) && (
              <Column className="px-10">
                <TextField
                  type="text"
                  value={optionObjects[0]?.label}
                  label=""
                  list="languageList"
                  valueCallback={(val) => languageSelectionCallBack?.(val)}
                  errors={languageErrorText()}
                ></TextField>

                <datalist id="languageList">
                  {languageOptions
                    ?.filter((item) => {
                      if (optionObjects.length > 0) {
                        return item.ncqaLanguageDesc
                          .toLowerCase()
                          .startsWith(optionObjects[0]?.label.toLowerCase());
                      } else {
                        return item;
                      }
                    })
                    .slice(0, 4)
                    .map((item) => (
                      <option
                        key={item.ncqaLanguageDesc}
                        value={item.ncqaLanguageDesc}
                        data-value={item.ncqaLanguageCode}
                      >
                        {item.ncqaLanguageDesc}
                      </option>
                    ))}
                </datalist>
              </Column>
            )}
            <Radio
              label="Decline to answer"
              selected={
                Boolean(isDeclineLanguageSelected) &&
                Boolean(languageEmptySelect)
              }
              value="Z2"
              callback={(val) => selectionCallBack?.(val)}
            />
          </div>
        )}
        {type === 'currentselection' && (
          <Column className=" body-1 px-1">
            {currentSelectionText(optionObjects)}

            <Spacer size={8} />
            <Column onClick={saveCallback}>
              <Title
                className="font-bold primary-color"
                text="Update"
                suffix={<Image src={editIcon} alt="link" />}
              />
            </Column>
          </Column>
        )}
      </Column>
      <Spacer size={16} />
      <Row className="flex-row">
        {type != 'currentselection' && (
          <Button
            className="w-[250px]"
            label="Save Answer"
            type="primary"
            callback={checkImpersonationForSaveCallback}
            disable={isImpersonated}
          />
        )}
        <Spacer axis="horizontal" size={16} />
        {type != 'currentselection' && optionsEnabled && (
          <Button
            className="w-[250px]"
            label="Cancel"
            type="secondary"
            callback={cancelCallback}
          />
        )}
      </Row>
      <Spacer size={32} />
      {divider && <Divider />}
    </Column>
  );
};
