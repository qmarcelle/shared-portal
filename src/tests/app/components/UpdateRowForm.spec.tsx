import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';

import { TextBox } from '@/components/foundation/TextBox';
import { UpdateRowForm } from '../../../app/otherProfileSettings/components/UpdateRowForm';

const renderUI = () => {
  const ethnicityList = [
    { label: 'Hispanic or Latino', enabled: true },
    { label: 'Not Hispanic or Latino', enabled: false },
    { label: 'Decline to answer', enabled: false },
  ];

  render(
    <UpdateRowForm
      label={
        <TextBox className="font-bold body-1" text="What is your ethnicity?" />
      }
      subLabel=""
      enabled={false}
      type={'radio'}
      optionObjects={ethnicityList}
      divider={true}
    />,
  );
};

describe('Race and Ethinicity slide', () => {
  it('should render the UI correctly', () => {
    const component = renderUI();
    const ethnicityList = [
      { label: 'Hispanic or Latino', enabled: true },
      { label: 'Not Hispanic or Latino', enabled: false },
      { label: 'Decline to answer', enabled: false },
    ];
    expect(screen.findByText('What is your ethnicity?'));
    ethnicityList.forEach((ethinicity) => {
      expect(screen.findAllByAltText(ethinicity.label)).toBeInTheDocument;
    });
    fireEvent.click(screen.getByRole('button', { name: /Save Answer/i }));
    expect(screen.findByText('What is your race?'));
    expect(screen.findByText(' How well do you speak English?'));

    expect(
      screen.findByText(
        'What language are you most comfortable using when speaking with your doctor?',
      ),
    );
    expect(screen.findByText('What language do you prefer to read?'));
    expect(component).toMatchSnapshot();
  });
});
