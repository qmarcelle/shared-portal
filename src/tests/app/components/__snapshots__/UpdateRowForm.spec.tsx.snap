// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Race and Ethinicity slide should render the UI correctly 1`] = `undefined`;
